High earning  in-app ads Analysts 

What the project is about: Do an in-depth study as to what characteristics and attribute yield the most profit.  
What your goal is in this project: Developed an application with the most successful attribute to increase the probability of success.  


To improve readability I will be adding function here with a short description.


In [994]:
# this allows the user to enter a CSV file name as a parameter to be used in a variable
from csv import reader
def extract_from_file(file_name):
    op=open(file_name)
    rf=reader(op)
    nlist=list(rf)
    return nlist[0],nlist[1:]
# this function will help explore the data  
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [995]:
android_header,android = extract_from_file('googleplaystore.csv')
ios_header, ios = extract_from_file('AppleStore.csv')


In [996]:
#printing the android data
print(android_header)
print('\n')
explore_data(android, 0, 2, True)


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


In [997]:
print(ios_header)
print('\n')
explore_data(ios, 0, 3, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


There is an error with a data set we were to find that error in a discussion thread. https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/discussion/106176 

By reading the discussion, we now know the issue is with the rating as there should not be a rating higher than 5.

With that information, I loop through the list to find any rating higher than 5.



In [998]:
cat=[]
at=[]
for count,row in enumerate(android):
    if float(row[2]) > 5: 
        at.append(count)
        cat.append(row[2])
print(cat)
print(at)
print(android[10472])

['19']
[10472]
['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


Which we know it to be at row 10472. Which is support by the following thread information

Source: https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps/discussion/106176 

state to be at index number 10472 or 10473 (depending on if you have header) has a missing value which can give an error


Per the instruction, we are to delete the row with data. Although I disagree with deleting the whole row as a data analysts what to preserve as much data as possible and would change the rating to 1.9 but to keep with the instruction I would go ahead and change the 19 to 1.9 and then delete. 

In [999]:
#changing 19 to 1.9
android[10472][2]=1.9
print(android[10472])


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', 1.9, '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [1000]:
# deleting that row
del android[10472]
print(android[10472])

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


Also in the discussion section discussed duplicates. Below is the code to find unique and duplicate entries using two lists and a for loop with the in operator.  

In [1001]:
duplicate_apps=[]
unique_apps=[]

for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)

print("Number of duplicate apps:", len(duplicate_apps))


Number of duplicate apps: 1181


As there is 1181 entries that are duplicates and below is show one instant of the app that multiple entries. 

In [1002]:
for app in android:
    name=app[0]
    if name == duplicate_apps[9]:
        print(app)

['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51507', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']
['Slack', 'BUSINESS', '4.4', '51510', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'August 2, 2018', 'Varies with device', 'Varies with device']


Upon inspection, we can see all the information is identical with the exception of the number of downloads. In order to clean our data and the most accurate information. We will delete all records except the record with the highest downloads. We could have done it by date if there was a date-time string but it only dates. 

Below is the code to find delete all the duplicate expects the highest value download.


In [1003]:
# the number entrires we should have at the end of the cleaning process
print('Expected length:', len(android)-len(duplicate_apps))

Expected length: 9659


In [1004]:
# first we are finding the higest rating for the each app
reviews_max={}

for app in android:
    name=app[0]
    n_reviews=float(app[3])
    if name in reviews_max and reviews_max[name]<n_reviews:
        reviews_max[name]=n_reviews
    if name not in reviews_max:
        reviews_max[name]=n_reviews
# I should have the same length of 9659
len(reviews_max)
#print(reviews_max)

9659

In [1005]:
# We are going to delete all the duplicate entries.
android_clean=[]
already_added=[]

for apps in android:
    name= apps[0]
    n_reviews=float(apps[3])
    #print(n_reviews)
    #print(reviews_max[name])
    if n_reviews == reviews_max[name] and  name not in already_added:
        android_clean.append(apps)
        already_added.append(name)
len(android_clean)
# I should get 9650
        
    

9659

The above code was a two-step processed to the first step was the find the highest value of reviews. Then we can compare the name and value in a loop to add only the name and the highest value to a list that eliminates any duplicates.

Upon further research, we found that there are characters that are Non-english which we are the only interesting in the Us market. First is to identify which app name has character other then English we will build a function to return false if a more then three-letter is grater the 127 otherwise return true.

In [1006]:
def Isenglish(astring):
    count=0
    for letter in astring:
        number_code = ord(letter)
        #print(number_code)
        if number_code >127 :
            count+=1
    if count > 3:
        return False
    else:
        return True
    
            
    

Test the fucntion Isenglish

In [1007]:
Isenglish('Instagram')

True

In [1008]:
Isenglish('爱奇艺PPS -《欢乐颂2》电视剧热播')

False

In [1009]:
Isenglish('Docs To Go™ Free Office Suite')

True

In [1010]:
Isenglish('Instachat 😜')

True

In [1011]:
USMarketAndroid=[]
for apps in android_clean:
    name=apps[0]
    if Isenglish(name)==1:
        USMarketAndroid.append(apps)
len(USMarket)


9614

In [1012]:
USMarketIos=[]
for apps in ios:
    name=apps[1]
    if Isenglish(name)==1:
        USMarketIos.append(apps)
len(USMarketIos)

6183

Now that we have isolated the US market for both platforms, we want to exclude paid applications and keep free applications in our data set, by exploring the data heading we see that price is an index at 7 for android and index 4 at for ios.

In [1013]:
print(android_header)
print('\n')

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']




In [1014]:
print(ios_header)
print('\n')

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']




In [1015]:
USMarketAndroidFree=[]
for apps in USMarketAndroid:
    price=apps[7]
    if price == '0':
        USMarketAndroidFree.append(apps)
len(USMarketAndroidFree)


    

8864

In [1016]:
USMarketIosFree=[]
for apps in USMarketIos:
    price=float(apps[4])
    if price == 0:
        USMarketIosFree.append(apps)
len(USMarketIosFree)


3222

Our end goal is to build a popular app for both market domain in Google Play and App Store.

The first step to accomplish this task is to build a simple premise of the ideal and implemented using a little resource, by adding it to Google Play.

The second step is to monitor the response and should show an upward trend to further develop the applications. 

The third step is proving the feasibility of the application by putting profit benchmarks and when the benchmarks are met to also deploy the application on the ios platform and submitting it to the App store.

Using the data set that we have on hand, we will explore and see what columns would be beneficial to generate a frequency table to uncover hidden trends.

We do this by print the head columns of both data sets.




In [1017]:
print(android_header)
print('\n')
print(ios_header)
print('\n')

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']




Now that we review the header from both markets we can use category from the google store and prime_genre from the app store as useful matrics in popularity.

We do this by making a frequency table functions, which take the argument of a list and index what column we what a frequency table for. Which would be for Andriod be index 1 and for ios would be index 11.


In [1018]:
def freq_table(a_list, a_index):
    return_dic={}
    for apps in a_list:
        name = apps[a_index]
        if name in return_dic:
            return_dic[name]+=1
        else:
            return_dic[name]=1
    total=len(a_list)
    for key,vlaue in return_dic.items():
        return_dic[key]=round((vlaue/total)*100,2)
        
    return return_dic

In [1019]:
# testing function for android data sets
android_test=freq_table(android,1)
print(android_test)

{'BUSINESS': 4.24, 'LIBRARIES_AND_DEMO': 0.78, 'BOOKS_AND_REFERENCE': 2.13, 'MAPS_AND_NAVIGATION': 1.26, 'PERSONALIZATION': 3.62, 'GAME': 10.55, 'FINANCE': 3.38, 'DATING': 2.16, 'HEALTH_AND_FITNESS': 3.15, 'TOOLS': 7.78, 'TRAVEL_AND_LOCAL': 2.38, 'BEAUTY': 0.49, 'ART_AND_DESIGN': 0.6, 'PHOTOGRAPHY': 3.09, 'SOCIAL': 2.72, 'HOUSE_AND_HOME': 0.81, 'AUTO_AND_VEHICLES': 0.78, 'FAMILY': 18.19, 'COMMUNICATION': 3.57, 'COMICS': 0.55, 'PARENTING': 0.55, 'MEDICAL': 4.27, 'VIDEO_PLAYERS': 1.61, 'WEATHER': 0.76, 'PRODUCTIVITY': 3.91, 'EVENTS': 0.59, 'SPORTS': 3.54, 'LIFESTYLE': 3.52, 'NEWS_AND_MAGAZINES': 2.61, 'FOOD_AND_DRINK': 1.17, 'EDUCATION': 1.44, 'ENTERTAINMENT': 1.37, 'SHOPPING': 2.4}


In [1020]:
# testing fuction for ios data sets
ios_test=freq_table(ios,11)
print(ios_test)

{'Utilities': 3.45, 'Business': 0.79, 'Book': 1.56, 'Navigation': 0.64, 'Reference': 0.89, 'Games': 53.66, 'Sports': 1.58, 'Health & Fitness': 2.5, 'Productivity': 2.47, 'Music': 1.92, 'Education': 6.29, 'Photo & Video': 4.85, 'Social Networking': 2.32, 'Finance': 1.45, 'Travel': 1.13, 'Weather': 1.0, 'Catalogs': 0.14, 'Food & Drink': 0.88, 'Shopping': 1.7, 'Entertainment': 7.43, 'News': 1.04, 'Medical': 0.32, 'Lifestyle': 2.0}


In [1021]:
# a function that will sort the frequancy table which include my freq_table function
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [1022]:
# display_table
display_table(USMarketIosFree,11)

Games : 58.16
Entertainment : 7.88
Photo & Video : 4.97
Education : 3.66
Social Networking : 3.29
Shopping : 2.61
Utilities : 2.51
Sports : 2.14
Music : 2.05
Health & Fitness : 2.02
Productivity : 1.74
Lifestyle : 1.58
News : 1.33
Travel : 1.24
Finance : 1.12
Weather : 0.87
Food & Drink : 0.81
Reference : 0.56
Business : 0.53
Book : 0.43
Navigation : 0.19
Medical : 0.19
Catalogs : 0.12


As we can see from the results  there some insight we can see such as:

The most common genre: Games
Second common genre: Entertainment

If we were to do a grouping among genre we can see that most downloads are entertainment-based rather then functional base

With this data set we can see there is an oversaturation in the entertainment section which would be harder to develop a break out of the app. But what we can focus on education app as it one of the none-Entertainment genre among the top 5 genres. If done with the approach of entertainment we can be a leader in an emerging genre. 


In [1023]:
display_table(USMarketAndroidFree,1)

FAMILY : 18.91
GAME : 9.72
TOOLS : 8.46
BUSINESS : 4.59
LIFESTYLE : 3.9
PRODUCTIVITY : 3.89
FINANCE : 3.7
MEDICAL : 3.53
SPORTS : 3.4
PERSONALIZATION : 3.32
COMMUNICATION : 3.24
HEALTH_AND_FITNESS : 3.08
PHOTOGRAPHY : 2.94
NEWS_AND_MAGAZINES : 2.8
SOCIAL : 2.66
TRAVEL_AND_LOCAL : 2.34
SHOPPING : 2.25
BOOKS_AND_REFERENCE : 2.14
DATING : 1.86
VIDEO_PLAYERS : 1.79
MAPS_AND_NAVIGATION : 1.4
FOOD_AND_DRINK : 1.24
EDUCATION : 1.16
ENTERTAINMENT : 0.96
LIBRARIES_AND_DEMO : 0.94
AUTO_AND_VEHICLES : 0.93
HOUSE_AND_HOME : 0.82
WEATHER : 0.8
EVENTS : 0.71
PARENTING : 0.65
ART_AND_DESIGN : 0.64
COMICS : 0.62
BEAUTY : 0.6


In [1024]:
display_table(USMarketAndroidFree,9)

Tools : 8.45
Entertainment : 6.07
Education : 5.35
Business : 4.59
Productivity : 3.89
Lifestyle : 3.89
Finance : 3.7
Medical : 3.53
Sports : 3.46
Personalization : 3.32
Communication : 3.24
Action : 3.1
Health & Fitness : 3.08
Photography : 2.94
News & Magazines : 2.8
Social : 2.66
Travel & Local : 2.32
Shopping : 2.25
Books & Reference : 2.14
Simulation : 2.04
Dating : 1.86
Arcade : 1.85
Video Players & Editors : 1.77
Casual : 1.76
Maps & Navigation : 1.4
Food & Drink : 1.24
Puzzle : 1.13
Racing : 0.99
Role Playing : 0.94
Libraries & Demo : 0.94
Auto & Vehicles : 0.93
Strategy : 0.91
House & Home : 0.82
Weather : 0.8
Events : 0.71
Adventure : 0.68
Comics : 0.61
Beauty : 0.6
Art & Design : 0.6
Parenting : 0.5
Card : 0.45
Casino : 0.43
Trivia : 0.42
Educational;Education : 0.39
Board : 0.38
Educational : 0.37
Education;Education : 0.34
Word : 0.26
Casual;Pretend Play : 0.24
Music : 0.2
Racing;Action & Adventure : 0.17
Puzzle;Brain Games : 0.17
Entertainment;Music & Video : 0.17
Casual;

Analyzing the frequency table Google play we can see that the top category are Family, Game, Tool Business and Genres would be Tools, Entertainment, education, and business.

So as we compare two data set we see that ios developers are focus on the entertainment while android developers are more balanced in having both practical apps and fun apps.  

We should also include the number of downloads to give better evidence of our analysis. 


To further support of analysis we are going to use the install column from the google play data set but since the app store does not have this column we will use the total number of user ratings as proximation to the total install.


In [1025]:
prime_genre =freq_table(USMarketIosFree,-5)

for genre in prime_genre:
    total=0
    len_genre=0
    for apps in USMarketIosFree:
        genre_app=apps[-5]
        if genre_app==genre:
            total+=float(apps[5])
            len_genre+=1
    avg=total/len_genre
    print(genre,":",avg)
    #print(genre)

Utilities : 18684.456790123455
Business : 7491.117647058823
Book : 39758.5
Navigation : 86090.33333333333
Reference : 74942.11111111111
Games : 22788.6696905016
Sports : 23008.898550724636
Health & Fitness : 23298.015384615384
Productivity : 21028.410714285714
Music : 57326.530303030304
Education : 7003.983050847458
Photo & Video : 28441.54375
Social Networking : 71548.34905660378
Finance : 31467.944444444445
Travel : 28243.8
Weather : 52279.892857142855
Catalogs : 4004.0
Food & Drink : 33333.92307692308
Shopping : 26919.690476190477
Entertainment : 14029.830708661417
News : 21248.023255813954
Medical : 612.0
Lifestyle : 16485.764705882353


Further diving into the data we do see that an app profile such as reference is the top applications with the most downloads, followed by weather, lifestyle. 
Which now we can focus on low app development in the market place with a high download count.


In [1026]:
googleCat=freq_table(USMarketAndroidFree,1)
for category in googleCat:
    total=0
    len_category=0
    for apps in USMarketAndroidFree:
        category_app=apps[1]
        if category_app==category:
            n_install=apps[5]
            n_install =n_install.replace('+','')
            n_install= n_install.replace(',','')
            total+=float(n_install)
            len_category+=1
    avg=total/len_category
    print(category,":",avg)

BUSINESS : 1712290.1474201474
LIBRARIES_AND_DEMO : 638503.734939759
BOOKS_AND_REFERENCE : 8767811.894736841
MAPS_AND_NAVIGATION : 4056941.7741935486
PERSONALIZATION : 5201482.6122448975
GAME : 15588015.603248259
FINANCE : 1387692.475609756
DATING : 854028.8303030303
HEALTH_AND_FITNESS : 4188821.9853479853
TOOLS : 10801391.298666667
TRAVEL_AND_LOCAL : 13984077.710144928
BEAUTY : 513151.88679245283
ART_AND_DESIGN : 1986335.0877192982
PHOTOGRAPHY : 17840110.40229885
SOCIAL : 23253652.127118643
HOUSE_AND_HOME : 1331540.5616438356
AUTO_AND_VEHICLES : 647317.8170731707
FAMILY : 3695641.8198090694
COMMUNICATION : 38456119.167247385
COMICS : 817657.2727272727
PARENTING : 542603.6206896552
MEDICAL : 120550.61980830671
VIDEO_PLAYERS : 24727872.452830188
WEATHER : 5074486.197183099
PRODUCTIVITY : 16787331.344927534
EVENTS : 253542.22222222222
SPORTS : 3638640.1428571427
LIFESTYLE : 1437816.2687861272
NEWS_AND_MAGAZINES : 9549178.467741935
FOOD_AND_DRINK : 1924897.7363636363
EDUCATION : 1833495.14

Looking at both data sets we can see that there room to grow in the education application for both google and ios. 